### PR TITLE
LiteLLM integration

### PR DESCRIPTION
**Pull request by galshubeli**
## Summary by CodeRabbit

- **New Features**
	- Introduced support for the LiteLLM model for chat completions.
	- Added a new Knowledge Graph testing framework for LiteLLM.
	- Expanded SDK capabilities to include support for Ollama and LiteLLM models.

- **Bug Fixes**
	- Enhanced error handling in the ontology creation process and JSON processing.

- **Documentation**
	- Updated README to reflect new model support and SDK capabilities.
	- Modified `.env.template` to include new environment variable declarations.

- **Chores**
	- Added optional dependency for LiteLLM in project configuration.
	- Updated wordlist with new entries.

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -26,13 +26,15 @@ vertexai = { version = "^1.49.0", optional = true }
 ollama = { version = "^0.2.1", optional = true }
 openai = { version = "^1.35.9", optional = true }
 google-generativeai = { version = "^0.8.1", optional = true }
+litellm = { version = "^1.52.14", optional = true }
```
[the remark this answers is not in the document]
Comment (gkorland) [reply pyproject.toml:29]: @galshubeli

File: tests/test_kg_litellm_openai.py [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+    def test_kg_creation(self):
+
+        file_path = "tests/data/madoff.txt"
+
+        sources = [Source(file_path)]
+
+        self.kg.process_sources(sources)
+
+        chat = self.kg.chat_session()
+        answer = chat.send_message("How many actors acted in a movie?")
```
Comment (swilly22) [inline tests/test_kg_litellm_openai.py:87]: this question is a bit too generic, which movie are we talking about ?

File: tests/test_kg_litellm_openai.py [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+        num_actors = 0 if len(actors_count) == 0 else int(actors_count[0])
+
+        assert num_actors > 10, "The number of actors found should be greater than 10"
+
+    def test_kg_delete(self):
+        self.kg.delete()
+
+        db = FalkorDB()
+        graphs = db.list_graphs()
+        self.assertNotIn(self.graph_name, graphs)
```
Comment (swilly22) [inline tests/test_kg_litellm_openai.py:102]: This doesn't seems like the right place for this test, it has nothing to do with litellm.

File: graphrag_sdk/models/litellm.py [hunk 4]
```diff
@@ -0,0 +1,283 @@
...
+        Args:
+            model_name (str): The name and the provider for the LiteLLM client.
+            generation_config (Optional[GenerativeModelConfig]): Configuration settings for generation.
+            system_instruction (Optional[str]): Instruction to guide the model.
+        """
+
+        self.model = model_name
+        if "/" in model_name:
+            self.provider = model_name.split("/")[0]
+            self.model_name = model_name.split("/")[1]
```
Comment (swilly22) [inline graphrag_sdk/models/litellm.py:41]: Add comment showing an example for a model name containing '/'

File: graphrag_sdk/models/litellm.py [hunk 5]
```diff
@@ -0,0 +1,283 @@
...
+        else:
+            self.provider = "openai"
+            self.model_name = model_name
```
Comment (swilly22) [inline graphrag_sdk/models/litellm.py:44]: A bit weird to assume `openai` in this case.

File: graphrag_sdk/models/litellm.py [hunk 6]
```diff
@@ -0,0 +1,283 @@
...
+            if not os.getenv("AZURE_API_KEY") or not os.getenv("AZURE_API_BASE") or not os.getenv("AZURE_API_VERSION"):
+                raise ValueError("Missing Azure-OpenAI credentials in the environment. See the Github Readme for more information.")
+        
+        elif self.provider == "gemini":
+            if not os.getenv("GEMINI_API_KEY"):
+                raise ValueError("Missing Gemini API key in the environment.")
+        
+        elif self.provider == "ollama":
+            self.ollama_client = Client()
+            self.check_and_pull_model()
```
Comment (swilly22) [inline graphrag_sdk/models/litellm.py:71]: are these all of the different LLM vendor litellm supports ?

File: graphrag_sdk/models/litellm.py [hunk 7]
```diff
@@ -0,0 +1,283 @@
...
+            
+    def check_and_pull_model(self) -> None:
```
Comment (swilly22) [inline graphrag_sdk/models/litellm.py:73]: Should we validate which provider is set ?
I guess this function shouldn't be called when `provider` is openai

File: graphrag_sdk/models/litellm.py [hunk 8]
```diff
@@ -0,0 +1,283 @@
...
+        """
+        return GenerationResponse(
+            text=response.choices[0].message.content,
+            finish_reason=(
+                FinishReason.STOP
+                if response.choices[0].finish_reason == "stop"
+                else (
+                    FinishReason.MAX_TOKENS
+                    if response.choices[0].finish_reason == "length"
+                    else FinishReason.OTHER
```
Comment (swilly22) [inline graphrag_sdk/models/litellm.py:142]: shouldn't there be a `FinishReason` which indicate success,
both `stop` and `max_tokens` indicate some sort of an error, `other` is unclear.

File: graphrag_sdk/models/litellm.py [hunk 9]
```diff
@@ -0,0 +1,283 @@
...
+            dict: The adjusted configuration settings for generation.
+        """
+        config = self._model.generation_config.to_json()
+        if output_method == OutputMethod.JSON:
+            config['temperature'] = 0
+            config['response_format'] = { "type": "json_object" }
+        
+        return config
+    
+    def delete_last_message(self):
```
Comment (swilly22) [inline graphrag_sdk/models/litellm.py:253]: Remind me, why do we need this function?

**Review by swilly22 — CHANGES_REQUESTED** [review]

File: graphrag_sdk/models/litellm.py (hunk 9 above)
Comment (galshubeli) [reply graphrag_sdk/models/litellm.py:253]: to clear bad Cypher generation from the history

File: graphrag_sdk/models/litellm.py (hunk 8 above)
Comment (galshubeli) [reply graphrag_sdk/models/litellm.py:142]: Currently we only track output errors, why do you think we need an indication of success?

File: graphrag_sdk/models/litellm.py (hunk 7 above)
Comment (galshubeli) [reply graphrag_sdk/models/litellm.py:73]: Line 70 is the end of the validation and this called only for Ollama provider

File: graphrag_sdk/models/litellm.py (hunk 6 above)
Comment (galshubeli) [reply graphrag_sdk/models/litellm.py:71]: There are more, but I noticed that they perform credentials validation poorly

File: graphrag_sdk/models/litellm.py (hunk 5 above)
Comment (galshubeli) [reply graphrag_sdk/models/litellm.py:44]: OpenAI is the only vendor that not need to specify the provider:
"azure/gpt-4o" -> azure
"gpt-4o"/"openai/gpt-4o" -> openai

File: graphrag_sdk/models/litellm.py (hunk 4 above)
Comment (galshubeli) [reply graphrag_sdk/models/litellm.py:46]: Done

File: tests/test_kg_litellm_openai.py (hunk 2 above)
Comment (galshubeli) [reply tests/test_kg_litellm_openai.py]: changed please review again

**Review by swilly22 — APPROVED** [review]

**Pull request merged**